### PR TITLE
Dvernon/postits

### PR DESCRIPTION
**Pull request by dvernon-tacc**

### CODE REVIEW

File: lib/src/main/java/edu/utexas/tacc/tapis/files/lib/caches/TenantAdminCache.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+
+public class TenantAdminCache extends BasePermsCache {
+    private final LoadingCache<TenantAdminCache.TenantAdminCacheKey, Boolean> cache;
+    private static final Logger log = LoggerFactory.getLogger(FilePermsCache.class);
+
+    public TenantAdminCache() {
+        cache = CacheBuilder.newBuilder().expireAfterWrite(Duration.ofSeconds(60)).build(new TenantAdminLoader());
```
Comment (dvernon-tacc) [inline lib/src/main/java/edu/utexas/tacc/tapis/files/lib/caches/TenantAdminCache.java]: I should probably make this longer than 60 seconds.  It's probably rare that someone's tenant admin status changes.  Maybe 5-10 mins?

File: lib/src/main/resources/edu/utexas/tacc/tapis/files/lib/FilesMessages.properties [hunk 2]
```diff
@@ -59,6 +60,30 @@ FILES_PERM_NOT_AUTH=FILES_PERM_NOT_AUTH User not authorized to modify permission
...
+POSTITS_DAO_ERROR_ID=POSTITS_DAO_ERROR_ID Database operation failed failed {0} {1}
+# 0 = ErrorMessage
+POSTITS_DAO_LIST_ERROR=POSTITS_DAO_LIST_ERROR Select failed {0} 
+# 0 = tenant, 1 = ApiUserId, 2 = oboTenant, 3 = oboUser, 4 = systemId, 5 = path
+POSTITS_REDEEM_DIR_NOZIP=POSTITS_REDEEM_DIR_NOZIP Must use zip when downloading a directory. Tenant: {0} ApiUserId: {1} OboTenant: {2} OboUser: {3} System: {4} Path: {5}
+# 0 = dbname, 1 = id, 2 = ErrorMessage
+POSTITS_ERR_NOT_REDEEMABLE=POSTITS_ERR_NOT_REDEEMABLE PostIt is not found or not redeemable {0} {1} {2}
+# 0 = postItId
+POSTIT_NOT_FOUND=POSTIT_NOT_FOUND PostIt not found {0}
+# 0 = postItId
```
Comment (dvernon-tacc) [inline lib/src/main/resources/edu/utexas/tacc/tapis/files/lib/FilesMessages.properties:84]: I need to remove this one message - it's a leftover from a failed experiment.

File: lib/src/main/java/edu/utexas/tacc/tapis/files/lib/caches/TenantAdminCache.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    public Boolean checkPerm(String tenantId, String username)
+            throws ServiceException
+    {
+        try {
+            TenantAdminCacheKey key = new TenantAdminCacheKey(tenantId, username);
+            return cache.get(key);
+        } catch (ExecutionException ex) {
+            // TODO:  Wrong message - correct this
+            String msg = LibUtils.getMsg("FILES_CACHE_ERR", "FilePerms", tenantId, username, ex.getMessage());
```
Comment (dvernon-tacc) [inline lib/src/main/java/edu/utexas/tacc/tapis/files/lib/caches/TenantAdminCache.java]: Just realized I still have a "TODO" to correct this message.  I'll take care of that.

**Comment by richcar58** [comment]
Here are some topics we could talk about later today.

1. Initial trace messages in PostitsResource
    - Since logging is our main resource during debugging, these messages should come as early as possible in the endpoint code.  Specifically, before code that can fail.

2. Runtime exceptions     
    - Do the messages generated from JAXRS exceptions (BadRequestException, ForbiddenException, etc.) provide users with a good explanation of the problem?

3. PostitsService:93-99 has magic numbers.  Let's talk about default values and whether they can be added to the json schema files.

4. UUIDs
    - Which UUID version is generated?  Routines.java contains several UUID generation methods, which one do we use.
    - Do we want postit UUIDs to have a Tapis suffix to make them easily identifiable?  See lines 17 and 38 in edu/utexas/tacc/tapis/shared/uuid/UUIDType.java in tapis-sharedlib and the TapisUUID class.

5. Should we also support POST or PUT as well as PATCH on the /{postitId} endpoint?

6. FILES_POSTITS table definition
    - Should there be an auto-seqenced field?
    - Assuming ID field is the postit UUID, should it be named UUID?
    - Why is clob used instead of varchar?
    - Are we saving all DateTimes in UTC?
    - Should jwtuser/jwttenant be creator/creatorTenant?
    - How is "never expires" represented in the table?
    - There seems to be only 2 indexes on the table, should there be more?

7. JOOQ discussion
    - Rationale for using it, pros and cons.
    - Could we include a README.jooq file in some obvious place that provices the precise instructions of when and how to regenerate the db files?
    
8. Let's walk through the cache code changes.

**Pull request merged**